src/common/IsolatedWorld.ts

```diff
@@ -507,74 +495,103 @@ export class IsolatedWorld {
       }
       debugError(error);
     }
-    function deliverResult(name: string, seq: number, result: unknown): void {
-      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-      // @ts-ignore Code is evaluated in a different context.
-      (globalThis as any)[name].callbacks.get(seq).resolve(result);
-      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-      // @ts-ignore Code is evaluated in a different context.
-      (globalThis as any)[name].callbacks.delete(seq);
-    }
   };
 
   async _waitForSelectorInPage(
     queryOne: Function,
     root: ElementHandle<Node> | undefined,
     selector: string,
     options: WaitForSelectorOptions,
-    binding?: PageBinding
+    bindings = new Set<(...args: never[]) => unknown>()
   ): Promise<JSHandle<unknown> | null> {
     const {
       visible: waitForVisible = false,
       hidden: waitForHidden = false,
       timeout = this.#timeoutSettings.timeout(),
     } = options;
-    const polling = waitForVisible || waitForHidden ? 'raf' : 'mutation';
-    const title = `selector \`${selector}\`${
-      waitForHidden ? ' to be hidden' : ''
-    }`;
-    async function predicate(
-      root: Element | Document,
-      selector: string,
-      waitForVisible: boolean,
-      waitForHidden: boolean
-    ): Promise<Node | null | boolean> {
-      const node = (await predicateQueryHandler(root, selector)) as Element;
-      return checkWaitForOptions(node, waitForVisible, waitForHidden);
+
+    try {
+      const handle = await this.waitForFunction(
+        async (PuppeteerUtil, query, selector, root, visible) => {
+          if (!PuppeteerUtil) {
+            return;
+          }
+          const node = (await PuppeteerUtil.createFunction(query)(
+            root || document,
+            selector
+          )) as Node | null;
+          return PuppeteerUtil.checkVisibility(node, visible);
+        },
+        {
+          bindings,
+          polling: waitForVisible || waitForHidden ? 'raf' : 'mutation',
+          root,
+          timeout,
+        },
+        new LazyArg(async () => {
+          try {
+            // In case CDP fails.
```

Comment:
what kind of failure could there be?

Reply:
`LazyArg`s should necessarily be CDP-agnostic because we don't want evaluate calls throwing back up lazy arg errors which themselves are evaluate errors. As an example, navigation can cause a failure here.

Reply:
What I mean if we should define expected errors (e.g., navigation as you said) and still throw on other errors? at least let's add an error logging so that we see it in the debug log?

Reply:
In this case, there is no possible other error besides CDP errors (which must all be ignore). It's possible if we change the injected code that it will throw due to that, but we have tests for that.